compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -94,17 +94,7 @@ impl<'a> Parser<'a> {
 
     /// Parses an expression, forcing tokens to be collected
     pub fn parse_expr_force_collect(&mut self) -> PResult<'a, P<Expr>> {
-        // If we have outer attributes, then the call to `collect_tokens_trailing_token`
-        // will be made for us.
-        if matches!(self.token.kind, TokenKind::Pound | TokenKind::DocComment(..)) {
-            self.parse_expr()
-        } else {
-            // If we don't have outer attributes, then we need to ensure
-            // that collection happens by using `collect_tokens_no_attrs`.
-            // Expression don't support custom inner attributes, so `parse_expr`
-            // will never try to collect tokens if we don't have outer attributes.
-            self.collect_tokens_no_attrs(|this| this.parse_expr())
-        }
+        self.collect_tokens_no_attrs(|this| this.parse_expr())
```

Comment:
Normally, this would be incorrect - `collect_tokens_no_attrs` was only supposed to be used when the target was guaranteed to not have any outer attribute - if outer attributes are present, we need to capture their tokens separately so that we can create a proper `ReplaceRange`.

However, we only actually need to know the attribute start position when we're in `capture_cg` mode, which never happens when we're parsing nonterminals (we flatten all nonterminals before entering `capture_cfg` mode). Can you add a `debug_assert!(!self.capture_cfg)` to `parse_expr_force_collect` to make sure this method doesn't get misused?

I may do some refactoring in a follow-up PR.

Reply:
The assert will fire because `parse_expr_force_collect` is indeed used with `capture_cfg == true`.
Here - https://github.com/rust-lang/rust/blob/master/compiler/rustc_builtin_macros/src/cfg_eval.rs#L194

(The result of `cfg_eval` should have tokens collected, especially with #85073.)

Reply:
I assumed that wrapping into `collect_tokens_no_attrs` is ok because there are two cases:
- There are active attributes, then token collection will be called once again in `parse_expr`, will set tokens for the returned expression, so the tokens collected by `collect_tokens_no_attrs` will simply be thrown away.
(I'm not sure what happens with range replacement in this case.)
- There are no active attributes, then `collect_tokens_no_attrs` is actually doing the right thing and it will set tokens for the returned expression. (I'm again not sure what will happen with range replacement in this case.)

Reply:
I think it will happen to do the right thing with replace ranges - we'll end up creating a `ReplaceRange` covering the entire expression (including outer attributes), as expected. Since we'll throw away the new tokens, the final captured tokens will *not* include the outer attributes, which will allow us to correctly perform cfg-expansion.

I forgot about the fact that we use force collection at the 'top level' of `cfg_eval`. It's unfortunate that we'll lose the asertion, but I don't think keeping it is worth the extra complexity of keeping track of the 'top-level' status on top of everything else.